### PR TITLE
DPE-36 Refactor unit tests for add_mongodb_repo

### PR DESCRIPTION
**Pull request by MiaAltieri**
PR for JIRA task [DPE-36](https://warthogs.atlassian.net/jira/software/c/projects/DPE/boards/390?modal=detail&selectedIssue=DPE-36) This PR updates tests for:
```
test_add_mongodb_org_repository_success
test_add_mongodb_org_repository_gpg_fail_leads_to_blocked
test_add_mongodb_org_repository_cant_create_list_file_blocks
test_add_mongodb_org_repository_already_added_skips
```

So that they now meet the following criteria: 
1. They should not mock the object being tested, only classes further down the dependency chain that are impractical to instantiate within the context of a unit test (e.g. a remote system that may not be available when running the test)

2. The test should not be checking for specific implementations but for the outcome of it.  So, if I am testing a method that computes the value of x in a quadratic equation, we don’t care if the implementation is based on trial and error or it uses a known formula.  What we care about is that the results are correct even when considering edge cases.  Therefore, if the implementation changes, the unit tests don’t need to change.

Tests for log messages were removed as they are not full crucial to the action we are testing, and they could change when someone refactors code. This means they might fall into category 2 where they test implementations rather than the outcome of the functions.

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -139,6 +139,37 @@
...
+1B3ZFJLxB1TlVT2Hjr+oigp0RY9W1FCIdO2uhb9GImpaJ1Y0ZZqUkt/d9D8U2wcw
+SW6/6WYeO7wAi/zlJ25hrBwhxS2+88gM6wJ1yL9yrM9v8JUb7Kq0rCGsEO5kqscV
+AmX90wsF2cZ6gHR53eGIDbAJK0MO5RHR73aQ4bpTivPnoTx4HTj5fyhW9z8yCSOe
+BlQABoFFqFvOS7KBxoyIS3pxlDetWOSc6yQrvA1CwxnkB81OHNmJfWAbNbEtZkLm
+xs2c8CIh2R81yi6HUzAaxyDH7mrThbwX3hUe/wsaD1koV91G6bDD4Xx3zpa9DG/O
+HyB98+e983gslg==
+=IQQF
+-----END PGP PUBLIC KEY BLOCK-----
+"""
+
```
Comment (marcoppenheimer) [inline tests/unit/test_charm.py:172]: I would personally dump this config (and the others above it) into a `tests/fixtures/mongo_pub_key.txt` and replace these lines with fixtures that can be referenced by the tests below. 

Something like:

```python
@pytest.fixture
def mongodb_pub_key():
    with open("fixtures/mongo_db_pub_key.txt") as f:
        key = f.read()
    
    return key

def test_something_awesome(mongodb_pub_key):
    assert SOMETHING == mongodb_pub_key
```

Comment (marcoppenheimer) [reply tests/unit/test_charm.py:172]: Saves headroom.

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
...
+        # preset values
+        repo_map.return_value = set()
         urlopen.side_effect = URLError("urlopen error")
-        with self.assertLogs("charm", "ERROR") as logs:
-            self.harness.charm._add_mongodb_org_repository()
-            self.assertIn("ERROR:charm:failed to get GPG key, reason:", "".join(logs.output))
+        self.harness.charm._add_mongodb_org_repository()
 
+        # verify we don't add repo when an exception occurs and that we enter blocked state
+        from_repo_line.assert_not_called()
```
Comment (quiet-ranger) [inline tests/unit/test_charm.py:429]: This and the next two lines are not really necessary and very much implementation dependent.  Just to illustrate, we could change the implementation so that the call to from_repo_line is made before urlopen, everything would still be correct but the unit test would fail.   I think here it is sufficient to check that the charm ended up in the BlockedStatus

**Review by zeljkoagic — APPROVED** [review]

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
...
-            with self.assertLogs("charm", "ERROR") as logs:
-                from_repo_line.side_effect = exception
-                self.harness.charm._add_mongodb_org_repository()
-                self.assertIn(log_message, "".join(logs.output))
+        for exception in exceptions:
+            from_repo_line.side_effect = exception
+            self.harness.charm._add_mongodb_org_repository()
 
+            # verify we don't add repo when an exception occurs and that we enter blocked state
+            import_key.assert_not_called()
```
Comment (quiet-ranger) [inline tests/unit/test_charm.py:451]: This and the following line are not needed for the same reason as mentioned before.  In fact, logically they could not be called because they need the output from from_repo_line and we have set up the test so that call fails.

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
...
     ):
-        repo_map.return_value = ["deb-https://repo.mongodb.org/apt/ubuntu-focal/mongodb-org/5.0"]
+        repo_map.return_value = {"deb-https://repo.mongodb.org/apt/ubuntu-focal/mongodb-org/5.0"}
+
+        # verify we don't add repo when we already have repo
         self.harness.charm._add_mongodb_org_repository()
-        urlopen.assert_called()
         from_repo_line.assert_not_called()
+        import_key.assert_not_called()
+        self.assertEqual(
```
Comment (quiet-ranger) [inline tests/unit/test_charm.py:470]: This assertion is totally unnecessary because we have mocked the value in this very method and we have simply repeated the string twice in the space of fewer than 10 lines of code.

This also made me think that, perhaps, a better way to implement _add_mongodb_org_repository() would be to parameterise what we are adding.  If we did this, then the unit test itself would be less brittle as it would not have to repeat exactly the same strings as used in the implementation.

**Review by quiet-ranger — CHANGES_REQUESTED** [review]

File: tests/unit/test_charm.py [hunk 5]
```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
...
-        log_messages = [
-            "ERROR:charm:failed to add repository, invalid source: invalid source message",
-            "ERROR:charm:failed to add repository: value message",
-        ]
-        for exception, log_message in zip(exceptions, log_messages):
-            with self.assertLogs("charm", "ERROR") as logs:
-                from_repo_line.side_effect = exception
-                self.harness.charm._add_mongodb_org_repository()
-                self.assertIn(log_message, "".join(logs.output))
+        for exception in exceptions:
```
Comment (marcoppenheimer) [inline tests/unit/test_charm.py:446]: What is the desired behavior of `_add_mongodb_org_repository()` when it errors with a different error to these?                                                                                                                                                                           I wonder if changing `src/charm.py` to this might help:                                                                                      

```python
try:
    line = "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/5.0 multiverse"
    repo.import_key(key)
    return
except (apt.InavlidSourceError, ValueError):
    logger.exception("failed to add repository due to XYZ reason")  # change XYZ
finally:
    self.unit.status = BlockedStatus("couldn't install MongoDB")
```

That way, in your test you can raise a generic exception, and assert `BlockedStatus`

**Review by marcoppenheimer — APPROVED** [review]
Looks good, some comments but nothing major!

File: tests/unit/test_charm.py (hunk 2 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:429]: I agree that `from_repo_line` isn't necessary here!

But I think we should still check that we do not add the key, since if we get an error and add the key we're doing something wrong by adding to the repo when we shouldn't be. (The purpose of the function `_add_mongodb_org_repository` is to add a key to the repository mapping. If we get an error **we should NOT add the key** in any implementation of the code)

Similarly if we get a URL error from ` key = urlopen("...")` we will **not** want to call import key since that could cause problems later down the line, so making sure this doesn't get called here is also necessary. Since in any implementation we would not want to add a key that was the result of a `urlopen` error. (This should be true for all possible implementations of this function)

File: tests/unit/test_charm.py (hunk 3 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:451]: I added the checks 
```
import_key.assert_not_called()
self.assertEqual(repo_map.return_value, set())
```
Because I in future implementations someone could restructure the code and we would want to be sure that if they had code failing from the call to `from_repo_line` that they their implementation still obeys the principle logic of "if `from_repo_line` fails do not `import_key` or `add` the repo.

File: tests/unit/test_charm.py (hunk 4 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:470]: I think its good to mock this function since we don't want to import gpg keys for each run of the unit test. This would mean that if a developer contributes to this project and runs the unit test they will be importing the key onto their machine.

But we do want to check that `_add_mongodb_org_repository` calls import key. So we should in fact `import_key.assert_not_called()`

File: tests/unit/test_charm.py (hunk 1 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:172]: FULLY agree! Thanks for pointing this out! :)

File: tests/unit/test_charm.py (hunk 4 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:470]: As for `_add_mongodb_org_repository()` what exactly would you want parameterised? I am open to changing this function and want to be on the same page as you. 😄  (although things like the URL, repo_name, and repo_line) are constant since this charm is MongoDB vs5

File: tests/unit/test_charm.py [hunk 6]
```diff
@@ -139,6 +139,37 @@
     "\n",
 ]
 
+MONGODB_PUB_KEY = """-----BEGIN PGP PUBLIC KEY BLOCK-----
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/unit/test_charm.py:142]: Yes! Lets do that, I agree!

File: tests/unit/test_charm.py [hunk 7]
```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
...
-        with self.assertLogs("charm", "ERROR") as logs:
-            self.harness.charm._add_mongodb_org_repository()
-            self.assertIn("ERROR:charm:failed to get GPG key, reason:", "".join(logs.output))
+        self.harness.charm._add_mongodb_org_repository()
 
+        # verify we don't add repo when an exception occurs and that we enter blocked state
+        from_repo_line.assert_not_called()
+        import_key.assert_not_called()
+        self.assertEqual(repo_map.return_value, set())
         self.assertEqual(self.harness.charm.unit.status, BlockedStatus("couldn't install MongoDB"))
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/unit/test_charm.py:432]: Yes, I'll do this! :D

File: tests/unit/test_charm.py [hunk 8]
```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
...
     @patch("charm.apt.DebianRepository.import_key")
     @patch("charm.apt.DebianRepository.from_repo_line")
-    @patch("charm.urlopen")
+    @patch("charm.apt.RepositoryMapping")
     def test_add_mongodb_org_repository_already_added_skips(
-        self, urlopen, from_repo_line, import_key, repo_map
+        self, repo_map, from_repo_line, import_key
     ):
-        repo_map.return_value = ["deb-https://repo.mongodb.org/apt/ubuntu-focal/mongodb-org/5.0"]
+        repo_map.return_value = {"deb-https://repo.mongodb.org/apt/ubuntu-focal/mongodb-org/5.0"}
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/unit/test_charm.py:464]: Sounds good to me :D

File: tests/unit/test_charm.py (hunk 5 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:446]: That makes sense! :D

File: tests/unit/data/mongodb_config_args.txt [hunk 9]
```diff
@@ -0,0 +1,73 @@
+net,
```
Comment (marcoppenheimer) [inline tests/unit/data/mongodb_config_args.txt:1]: Suggest removing all the commas from end of lines.
See comment on the test for context.

File: tests/unit/test_charm.py [hunk 10]
```diff
@@ -712,3 +604,21 @@ def test_get_primary_peer_unit_primary(self, single_replica):
...
+    @property
+    def mongodb_config_args(self) -> List[str]:
+        with open("tests/unit/data/mongodb_config_args.txt") as f:
+            config_args = f.read()
+            config_args = config_args.split(",\n")
+            # handle reformatting issues
+            config_args = ["\n" if args == "\\n" else args for args in config_args]
+            config_args = config_args[:-1]
+
+        return config_args
```
Comment (marcoppenheimer) [inline tests/unit/test_charm.py:617]:
```suggestion
            config_args = f.read().splitlines()
            config_args_formatted = ["\n" if arg == "\\n" else arg for arg in config_args]
            
            return config_args
```

Comment (marcoppenheimer) [reply tests/unit/test_charm.py:617]: Related to my other comment about commas.

File: tests/unit/test_charm.py (hunk 5 above)
Comment (marcoppenheimer) [reply tests/unit/test_charm.py:298]: @MiaAltieri - Juan caught a bug in my suggestion, I edited the above suggestion for when you want it.

File: tests/unit/test_charm.py (hunk 4 above)
Comment (quiet-ranger) [reply tests/unit/test_charm.py:318]: This case is the same as before.  Most of the objects are temporary and only exist in memory.  They are needed for the following call to work:
`                repositories.add(repo)`
This is the real one that would cause a change in the system but, in this case, you have mocked it.  So, all we need is to assert that this add was not called in this test.  If this is still not clear, let's talk about it.

As for the parameterisation, it is not strictly mandatory for now but we discussed the advantages when thinking about the future with both Mongo and Percona and perhaps more than one version for each.

File: tests/unit/data/mongodb_config.txt [hunk 11]
```diff
@@ -0,0 +1,42 @@
+# mongod.conf
+
+# for documentation of all options, see:
+#   http://docs.mongodb.org/manual/reference/configuration-options/
+
+# Where and how to store data.
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/unit/data/mongodb_config.txt:6]: messaging you on MM to discuss consistency with default MongoDB conf files :)

File: tests/unit/test_charm.py (hunk 5 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:298]: Yeah! I saw that with the `finally` as well, I think we aren't supposed to keep the blank `except` either so I think the final form should be:
```
try:
    line = "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/5.0 multiverse"
    repo.import_key(key)
    return
except (apt.InavlidSourceError, ValueError):
    logger.exception("failed to add repository due to XYZ reason")  # change XYZ
    self.unit.status = BlockedStatus("couldn't install MongoDB")
```
Thanks for the help with this! I'll tag you in the next PR :)

**Review by paulomach — COMMENTED** [review]
@MiaAltieri, good code with decent coverage. Will wait until you figure out the last details with JJ though.

**Review by zeljkoagic — APPROVED** [review]
LGTM, minor changes only!

File: src/charm.py [hunk 12]
```diff
@@ -103,8 +103,14 @@ def _on_install(self, _) -> None:
         installs MongoDB.
         """
         self.unit.status = MaintenanceStatus("installing MongoDB")
-        self._add_mongodb_org_repository()
-        self._install_apt_packages(["mongodb-org"])
+        repo_name = "deb-https://repo.mongodb.org/apt/ubuntu-focal/mongodb-org/5.0"
+        repo_entry = "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/5.0 multiverse"
+        gpg_url = "https://www.mongodb.org/static/pgp/server-5.0.asc"
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py:108]: Would echo @marcoppenheimer here about both points.

Also, since AFAIK we should support a number of os releases (focal, and others), these could be retrieved from the metadata.

File: src/charm.py [hunk 13]
```diff
@@ -103,8 +103,14 @@ def _on_install(self, _) -> None:
...
+        try:
+            self._add_repository(repo_name, gpg_url, repo_entry)
+            self._install_apt_packages(["mongodb-org"])
+        except (apt.InvalidSourceError, ValueError, apt.GPGKeyError, URLError):
+            self.unit.status = BlockedStatus("couldn't install MongoDB")
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py:113]: I would take the opposite route as @marcoppenheimer , and be more granular on the messaging for the `BlockedStatus` , mainly due to bad experiences on not knowing the correct reason for a blockedstatus. 
Something among the lines of:

```suggestion
        except apt.InvalidSourceError:
            status_message = "Invalid Source.",
        except ValueError:
            status_message = "something descriptive or the error"
        ...
        except:
           status_message = ""
        finally:
            self.unit.status = BlockedStatus("Couldn't install MongoDB." + status_message )
```

[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:113]: A 'blank except' I think would mean just a try:except alone. Though, I'm not sure what the desired output would be for unexpected exceptions, would we want it to fail out? If you add a 'raise NotImplementedErorr' on line 109, what happens to the charm on deployment? Does it just break, or does something useful happen?

Either way, I think your current implementation @MiaAltieri is good .

File: src/charm.py [hunk 14]
```diff
@@ -218,35 +224,39 @@ def _open_port_tcp(self, port: int) -> None:
...
 
-        # Get GPG key
-        try:
-            key = urlopen("https://www.mongodb.org/static/pgp/server-5.0.asc").read().decode()
-        except URLError as e:
-            logger.exception("failed to get GPG key, reason: %s", e)
-            self.unit.status = BlockedStatus("couldn't install MongoDB")
-            return
+        Args:
+            repo_name: deb-https url of repo to add
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:233]: agreed!

File: src/charm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:108]: fully agree!

[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:108]: As for updating version, I'll handle that in a separate PR in the future when it comes times to deal with updating and versioning.
